libminifi/src/core/extension/ExtensionManager.cpp

```diff
@@ -0,0 +1,134 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/extension/ExtensionManager.h"
+#include "core/logging/LoggerConfiguration.h"
+#include "utils/file/FileUtils.h"
+#include "core/extension/Executable.h"
+#include "utils/file/FileMatcher.h"
+#include "core/extension/DynamicLibrary.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace core {
+namespace extension {
+
+namespace {
+struct LibraryDescriptor {
+  std::string name;
+  std::string dir;
+  std::string filename;
+
+  bool verify(const std::shared_ptr<logging::Logger>& /*logger*/) const {
+    // TODO(adebreceni): check signature
+    return true;
+  }
```

Comment:
I wouldn't call a library descriptor "invalid" if it fails our (future) verification logic (a SignedExtensionLibrary would be such a class), I generally like to avoid throwing exceptions (especially from constructors) if possible (I only try to use them to cut short deep computations where checking the return value at each step would be cumbersome)

Reply:
The only was to signal an error from constructors is through exceptions, so it's hard to avoid them there. But in general I think exceptions are a great way to signal error.
Enforcing verification criteria in the constructor would allow to have the result of future verification be a class invariant, thus reducing the class state.

My exception usage preferences are quite the opposite of yours: [Using exceptions for errors only](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e3-use-exceptions-for-error-handling-only), [including establishing invariants](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e5-let-a-constructor-establish-an-invariant-and-throw-if-it-cannot), but not part of the happy path.

Reply:
making "verified" seem like a class invariant would be a mistake IMO, as the moment verification finishes somebody could overwrite the file on the filesystem rendering the invariant invalid, but as we do not provide any verification logic we could circle back to this question when we actually implement it

Reply:
I would be fine with a requirement that extension changes on disk must happen while the extensions are not loaded in minifi. They only need to change during minifi updates and possibly through C2 or registry fetches in the future. I consider them internals, so if someone randomly changes them, they are shooting themselves in the foot anyway. Just like you don't overwrite system dlls on windows while it's running.

Treating them like this would make them conceptually "const" and fine to base invariants on.